packages/hardhat-viem/src/internal/chains.ts

```diff
@@ -0,0 +1,94 @@
+import type { EthereumProvider } from "hardhat/types";
+import type { Chain } from "viem";
+import type { TestClientMode } from "viem/src/clients/createTestClient";
+
+import memoize from "lodash.memoize";
+
+import {
+  UnknownDevelopmentNetworkError,
+  NetworkNotFoundError,
+  MultipleMatchingNetworksError,
+} from "./errors";
+
+export async function getChain(provider: EthereumProvider): Promise<Chain> {
+  const chains: Record<string, Chain> = require("viem/chains");
+  const chainId = await getChainId(provider);
+
+  if (isDevelopmentNetwork(chainId)) {
+    if (await isHardhatNetwork(provider)) {
+      return chains.hardhat;
+    } else if (await isFoundryNetwork(provider)) {
+      return chains.foundry;
+    } else {
+      throw new UnknownDevelopmentNetworkError();
+    }
+  }
+
+  const matchingChains = Object.values(chains).filter(
+    ({ id }) => id === chainId
+  );
+
+  if (matchingChains.length === 0) {
+    throw new NetworkNotFoundError(chainId);
+  }
+
+  if (matchingChains.length > 1) {
+    throw new MultipleMatchingNetworksError(chainId);
+  }
+
+  return matchingChains[0];
+}
+
+export function isDevelopmentNetwork(chainId: number) {
+  return chainId === 31337;
+}
+
+export async function getMode(
+  provider: EthereumProvider
+): Promise<TestClientMode> {
+  if (await isHardhatNetwork(provider)) {
+    return "hardhat";
+  } else if (await isFoundryNetwork(provider)) {
+    return "anvil";
+  } else {
+    throw new UnknownDevelopmentNetworkError();
+  }
+}
+
+async function getChainId(provider: EthereumProvider) {
+  return _memoizedGetChainId(provider);
+}
+
+const _memoizedGetChainId = memoize(async (provider: EthereumProvider) => {
+  return Number(await provider.send("eth_chainId"));
+});
+
+async function isHardhatNetwork(provider: EthereumProvider) {
+  return _memoizedIsHardhatNetwork(provider);
+}
+
+const _memoizedIsHardhatNetwork = memoize(
```

Comment:
Just a personal preference:

What about declare a single method that handle all the scenarios? (especially if we expect more values in the future(?))

```
async function isNetworkWithProvider(
  provider: EthereumProvider,
  methodName: string
) {
  try {
    await provider.send(methodName);
    return true;
  } catch {
    return false;
  }
}
```

And then use it like this:

```
const _memoizedIsHardhatNetwork = memoize(
  async (provider: EthereumProvider) => {
    return isNetworkWithProvider(provider, "hardhat_metadata");
  }
```

"hardhat_metadata" and "anvil_nodeInfo" could be enums too